watch/vm: Handle VM generation updates

Acedus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
#### Before this PR:
Since the introduction of VM Controller Revisions, they only ever got created during VM startup. No new revisions were being created when a live-change was applied to the VM which made it so certain mechanisms which rely on the revision such as VM snapshot would make use of an outdated manifest.

Additionally, update logic of the VM's `DesiredGeneration` and `ObservedGeneration` relied on a VMI annotation which was being populated according to the VM revision which made the whole process more convoluted than it needed to be.

#### After this PR:
This PR introduces new logic to create new VM Controller Revisions according to the following set of factors:
- The VM is starting for the first time.
- The VM's `DesiredGeneration` differs from its `ObservedGeneration`.
- The VM lacks the `RestartRequired` condition.
- The template spec has changed.

Note that with this change **only** changes to the VM's template spec will trigger the creation of a new revision as only such changes affect the guest. With that said, other fields that may be desired such as `RunStrategy` can be considered for inclusion. 

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
 
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->
- Fixes https://issues.redhat.com/browse/CNV-61279
- Fixes https://issues.redhat.com/browse/CNV-61741

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:
- Creating revisions per change with a hard-cap on revision history that's either hard-coded or user configurable - Compliant with how `ControllerRevisions` work but extends beyond the scope of this fix, could be considered in a follow-up PR.

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer
- The first commit is required for this PR to work as it solves a race condition between `stopVMI`'s cleanup and subsequent revision creation by the new logic. Relevant PR: https://github.com/kubevirt/kubevirt/pull/15287

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
bugfix: Online snapshots now correctly include live-changes applied to a VM.
```

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/3321b5fe696de6f6cabbf78285aa4973196fa0ae..e5bfb28641da6b98a5236be07b04380127784b3d) - Addressed @lyarwood and @0xFelix comment.

Acedus commented:
/test pull-kubevirt-unit-test

0xFelix commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -3474,3 +3233,97 @@ func (c *Controller) handleDeclarativeVolumeHotplug(vm *virtv1.VirtualMachine, v
+}
+
+func shouldCreateRevision(vmSpec, revisionSpec virtv1.VirtualMachineSpec) bool {
+	vmSpec.Running = revisionSpec.Running
```
Can you explain why it is required? What about `RunStrategy`?

Acedus replied:
This and the following line are just to effectively ignore those fields during comparison. I figured it'd be better than to run `DeepEqual` for each and every field we wanted to take into account.

0xFelix replied:
And what about `RunStrategy`?

Acedus replied:
I don't think we should exclude `RunStrategy` as it technically is a field that is able to be live-updated.
I understand that setting to `Halted` for example would shut the VM down and still create a revision hence it isn't necessarily relevant (it's redundant to create a revision in that case) but that would change back to whatever the user specifies next that brings it back online. Scenarios like `Always` -> `RerunOnFailure` are definitely valid.

0xFelix replied:
I understand. Can it try to patch a VMI even though the strategy was changed to `Halted`?

Acedus replied:
So that's a fair point, and it will fail, but it should be eventually consistent as once the VMI starts again it should pick up the new revision anyway. I don't mind adding some logic to make sure that it wasn't changed to an invalid/redundant option.

0xFelix replied:
If we know in advance IMO we should avoid failures. It could produce misleading logs.

0xFelix replied:
Yeah, I'd like to see a `should not update generation and sync during Execute()`

Acedus commented:
/retest-required

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/09845ea6bcc2b396fe56c2240eaace8ca18a4428..66daa59c6e6f66e9adc263bd9c5a763898de2c49) - Address @0xFelix comment, also added some evaluation logic to `shouldCreateRevision` to create a new revision for swap between `running` and `runStrategy` and vice-versa.

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -2113,109 +2030,190 @@ var _ = Describe("VirtualMachine", func() {
+						},
+					),
+					Entry(
+						"with change from Running to RunStrategy",
```
Might be a silly one, but the inverse is cheap to test as well

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -2113,109 +2030,190 @@ var _ = Describe("VirtualMachine", func() {
+				)
+
+				DescribeTable("should not update generation and sync during Execute()", func(spec, newSpec v1.VirtualMachineSpec) {
+					const desiredGeneration = 3
+					const observedGeneration = 2
+					testutils.UpdateFakeKubeVirtClusterConfig(kvStore, &v1.KubeVirt{
+						Spec: v1.KubeVirtSpec{
+							Configuration: v1.KubeVirtConfiguration{
+								VMRolloutStrategy: pointer.P(v1.VMRolloutStrategyLiveUpdate),
+							},
+							WorkloadUpdateStrategy: v1.KubeVirtWorkloadUpdateStrategy{
+								WorkloadUpdateMethods: []v1.WorkloadUpdateMethod{
+									v1.WorkloadUpdateMethodLiveMigrate,
+								},
+							},
+						},
+					})
+
+					vm.Generation = 2
+					vm.Spec = spec
+
+					cr := createVMRevision(vm)
+					controller.crIndexer.Add(cr)
+
+					vmi.Status.VirtualMachineRevisionName = cr.Name
+
+					vm.Generation = 3
+					vm.Spec = newSpec
+
+					vm, err := virtFakeClient.KubevirtV1().VirtualMachines(vm.Namespace).Create(context.TODO(), vm, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					addVirtualMachine(vm)
+
+					vmi, err = virtFakeClient.KubevirtV1().VirtualMachineInstances(vm.Namespace).Create(context.TODO(), vmi, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					controller.vmiIndexer.Add(vmi)
+
+					sanityExecute(vm)
```
Can you dedup some of this into a `BeforeEach` in this `Context`?

Acedus replied:
Any specific suggestions? I'll move the generation assignment as all tests have it in common but because we need to create the revision with the initial spec I can't really move much more into a `BeforeEach`.

0xFelix replied:
By dedup I meant move some of the repeated setup code into a helper func.

orelmisan commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -1312,9 +1312,6 @@ const (
 	AllowPodBridgeNetworkLiveMigrationAnnotation string = "kubevirt.io/allow-pod-bridge-network-live-migration"
 
-	// VirtualMachineGenerationAnnotation is the generation of a Virtual Machine.
-	VirtualMachineGenerationAnnotation string = "kubevirt.io/vm-generation"
```
Will this annotation linger forever on existing VMI objects?

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/46338a2b6496b6743c4d5a1eb8cf434cf7a4337d..9d3db1bf5b6a5f425bca1f7e12e3daa0e0cc8230) - Address @0xFelix's comment.

Acedus commented:
> Looks good now I'd say. Do we have to wait for #15287 now?

Yes.

lyarwood reviewed (commented):
Looks pretty good overall, can you also add some test coverage for the specific bugs this work addresses, for example creating a VM from a snapshot that itself was taken after the original VM was live updated with more resources etc.

Acedus commented:
Added relevant snapshot tests and removed previous online snapshot logic that was responsible for altering the revision to include missing live-updated fields.

0xFelix commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -3435,3 +3239,116 @@ func (c *Controller) handleDeclarativeVolumeHotplug(vm *virtv1.VirtualMachine, v
+		return vmi, err
+	}
+
+	vmi, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
```
Side comment: Why do we have PatchStatus for VM but not for VMI?

0xFelix commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -3004,8 +2842,15 @@ func (c *Controller) syncVMAnnotationsToVMI(vm *virtv1.VirtualMachine, vmi *virt
 		descheduler.EvictPodAnnotationKeyAlphaPreferNoEviction,
 	}
 
+	removedDeprecated := false
```
nit: if super-lazy you could reuse `changed` for both for-loops that are essentially checking the changed condition

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -1983,114 +1997,282 @@ var _ = Describe("VirtualMachine", func() {
+					Expect(vmi.Status.VirtualMachineRevisionName).To(Equal(getVMRevisionName(vm.UID, vm.Status.ObservedGeneration)), "VMI status VM revision name should be")
+				})
+
+				It("should not bump observed generation nor create a new VM controller revision for run strategy change to halted", func() {
```
There are more conditions in `shouldCreateRevision`, are they all covered?

0xFelix commented on tests/storage/snapshot.go:
```diff
@@ -845,6 +846,204 @@ var _ = Describe(SIG("VirtualMachineSnapshot Tests", func() {
+						time.Minute)
+				})
+
+				It("should include cpu hotplug in vm snapshot", func() {
```
This and the following test could be combined into a single case to save resources? These tests look rather expensive.

0xFelix commented on pkg/storage/snapshot/snapshot_test.go:
```diff
@@ -1048,63 +1039,6 @@ var _ = Describe("Snapshot controlleer", func() {
 				Expect(*createCalls).To(Equal(1))
 			})
 
-			It("should create online VirtualMachineSnapshotContent with volume migration", func() {
```
This was replaced by a functional test?

Acedus replied:
Unfortunately yes, the conditional revision creation logic is the domain of a separate controller, and this interaction between the two justifies the transition to a functional test.

Acedus commented:
Addressed @0xFelix's comments.

Acedus commented:
@lyarwood is there anything else in your opinion that's missing from this PR?

Acedus commented:
/retest-required

Acedus commented:
/cc @enp0s3

Acedus commented:
/hold

Failures seem relevant.

0xFelix commented:
Unsure, could also be a flake / indeterministic behavior in VM pools. Please have a look.

Acedus commented:
/retest-required

Sreeja1725 commented:
> @Sreeja1725 @rthallisey Can you help out with this? What should be desired behavior for VM pools?

I don't think it's a desired behavior, it's a bug on the vmpool side, currently it is resetting or changing the immutable fields which must remain stable for the VM's lifetime. Will raise a fix to preserve the immutable fields.
cc: @rthallisey 

@Acedus Can this PR wait until the fix has been raised and merged?

Acedus commented:
Should be good now, @0xFelix @xpivarc.

Acedus commented:
/unhold

mhenriks commented:
@Acedus I just discovered this PR in attempting to help address #17229. I am not sure I agree with the approach here! 

Why have ControllerRevisions that are just copies of the current VM?

Not sure about stopping creating new ControllerRevisions when RestartRequired exists either. Volume hotplug still happens when RestartRequired exists

~~Another minor issue...the current code will include ephemal hotplug volumes in a snapshot. We didn't do that before on purpose~~

If this gets backported to v1.8, pci topology discovery in https://github.com/kubevirt/kubevirt/pull/17058 will break

mhenriks commented:
cc @akalenyu @awels

Acedus commented:
> @Acedus I just discovered this PR in attempting to help address #17229. I am not sure I agree with the approach here!
> 
> Why have ControllerRevisions that are just copies of the current VM?
> 
> Not sure about stopping creating new ControllerRevisions when RestartRequired exists either. Volume hotplug still happens when RestartRequired exists
> 
> ~Another minor issue...the current code will include ephemal hotplug volumes in a snapshot. We didn't do that before on purpose~
> 
> If this gets backported to v1.8, pci topology discovery in #17058 will break

Hi @mhenriks, you raise fair points. Thinking it through, I was naive about the invariants other subsystems rely on wrt the `ControllerRevision`... I think the right path forward is to revert this PR and follow up with:

1. Keep the logic that transitions from the VMI annotation to using `ObservedGeneration` since it's a cheap simplification that stands on its own regardless of revision lifecycle.
2. Re-fix the snapshot issue in a targeted way rather than updating the revision on every live change.
3. We should debate whether the way we utilize ControllerRevisions today is correct since I do believe there's a case to be made for including new(er) revisions on persistent VM live-changes.

mhenriks commented:
> > @Acedus I just discovered this PR in attempting to help address #17229. I am not sure I agree with the approach here!
> > Why have ControllerRevisions that are just copies of the current VM?
> > Not sure about stopping creating new ControllerRevisions when RestartRequired exists either. Volume hotplug still happens when RestartRequired exists
> > ~Another minor issue...the current code will include ephemal hotplug volumes in a snapshot. We didn't do that before on purpose~
> > If this gets backported to v1.8, pci topology discovery in #17058 will break
> 
> Hi @mhenriks, you raise fair points. Thinking it through, I was naive about the invariants other subsystems rely on wrt the `ControllerRevision`... I think the right path forward is to revert this PR and follow up with:
> 
> 1. Keep the logic that transitions from the VMI annotation to using `ObservedGeneration` since it's a cheap simplification that stands on its own regardless of revision lifecycle.
> 2. Re-fix the snapshot issue in a targeted way rather than updating the revision on every live change.
> 3. We should debate whether the way we utilize ControllerRevisions today is correct since I do believe there's a case to be made for including new(er) revisions on persistent VM live-changes.

Thanks @Acedus I know that ControllerRevisions were added to support snapshots but they are also very valuable in other cases where we need to know the spec of the vm at startup. Fixing #17229 without adding a new api field would be pretty complicated without controller revisions. I don't think #17058 would be possible without.

I agree with you on all three points above.

WRT 2 I think for `vm.spec.template.spec` we can use the current VMI spec MINUS any field that is NOT in the current VM spec OR the ControllerRevision. Don't want to store any of the VMI fields that get defaulted or weren't in the VM at some point
